src/core/nativecall.c

```diff
@@ -299,6 +299,13 @@ char * MVM_nativecall_unmarshal_string(MVMThreadContext *tc, MVMObject *value, M
                 str = MVM_string_utf8_encode_C_string(tc, value_str);
         }
 
+#ifdef MVM_USE_MIMALLOC
+        size_t str_len = strlen(str) + 1;
+        char *libc_str = malloc(str_len);
+        memcpy(libc_str, str, str_len);
+        MVM_free(str);
+#endif
```

Comment:
The changes in this file are the only part of the entire commit that is really bugging me, because the `strlen` is wrong for the case of `MVM_NATIVECALL_ARG_UTF16STR`

(It's also clear that we don't have any tests for this, but that's a fix needed in the rakudo testsuite, I guess, and shouldn't distract us here...)

What's also bothering me is that I'd like not to introduce a (new) bug. The only sensible way I can see to get the length of the UTF-16 string is to call `MVM_string_utf16_encode_substr` and pass in `&output_size`. But once I start down that path, it feels silly to use that approach for UTF-16, but `strlen` for the 8-bit variants. Which ultimately ends up with my minimal change being "less than tiny":

```
$ git diff --stat src/core/nativecall.c
 src/core/nativecall.c | 57 +++++++++++++++++++++++++++++++++------------------
 1 file changed, 37 insertions(+), 20 deletions(-)
```

and it's then easier to just paste the entire (revised) function:

```
char * MVM_nativecall_unmarshal_string(MVMThreadContext *tc, MVMObject *value, MVMint16 type, MVMint16 *free, MVMint16 unmarshal_kind) {
    if (IS_CONCRETE(value)) {
        /* Set whether to free the encoded string. */
        if (free) {
            if (REPR(value)->ID == MVM_REPR_ID_MVMCStr)
                *free = 0; /* Manually managed. */
            else if (free && type & MVM_NATIVECALL_ARG_FREE_STR_MASK)
                *free = 1;
            else
                *free = 0;
        }

        /* Encode string. */
        MVMString *value_str = MVM_repr_get_str(tc, value);
        MVMuint64 output_size;
        char *str;
        type &= MVM_NATIVECALL_ARG_TYPE_MASK;
        switch (type) {
            case MVM_NATIVECALL_ARG_ASCIISTR:
                str = MVM_string_ascii_encode(tc, value_str, &output_size, 0);
                break;
            case MVM_NATIVECALL_ARG_UTF16STR:
                str = MVM_string_utf16_encode_substr(tc, value_str, &output_size, 0, -1, NULL, 0);
                break;
            default:
#ifdef MVM_USE_MIMALLOC
                /* This *doesn't* write a NUL */
                str = MVM_string_utf8_encode(tc, value_str, &output_size, 0);
#else
                /* whilst this wrapper that calls it reallocs and adds a NUL */
                str = MVM_string_utf8_encode_C_string(tc, value_str);
#endif
        }

#ifdef MVM_USE_MIMALLOC
        char *libc_str;
        if (type == MVM_NATIVECALL_ARG_UTF16STR) {
            /* output_size is in 2 byte units, and doesn't include the
             * terminating U-0000, which was written. */
            size_t octet_len = 2 * (output_size + 1);
            libc_str = malloc(octet_len);
            memcpy(libc_str, str, octet_len);
        }
        else {
            /* This code is simpler if we ignore the detail that ...ascii_encode
             * wrote a NUL, but ...utf8_encode did not. */
            size_t octet_len = output_size + 1;
            libc_str = malloc(octet_len);
            memcpy(libc_str, str, output_size);
            libc_str[output_size] = 0;
        }
        MVM_free(str);
        return libc_str;
#else
        return str;
#endif
    }
    else {
        return NULL;
    }
}
```

"Ultimately" (ie next week if not sooner) I'd like to go with something like the functions I proposed in my branch to remove the repeated malloc, but I'd like to get this fix merged, as it fixes a regression from the previous release.

So (I think) on the basis that I don't like commits that knowingly break stuff (only to be fixed in the next commit) (because `git bisect` has no idea about this), I'd like **something** merged into this commit that avoids the new bug, but I'm not sure that I'm recommending my own suggestion in this comment.